compiler/rustc_typeck/src/check/upvar.rs

```diff
@@ -1053,6 +1053,51 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
     }
 }
 
+/// Truncate the capture so that the place being borrowed is in accordance with RFC 1240,
+/// which states that it's unsafe to take a reference into a struct marked `repr(packed)`.
+fn restrict_repr_packed_field_ref_capture<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    param_env: ty::ParamEnv<'tcx>,
+    place: &Place<'tcx>,
+) -> Place<'tcx> {
+    let pos = place.projections.iter().enumerate().position(|(i, p)| {
+        let ty = place.ty_before_projection(i);
+
+        match p.kind {
+            ProjectionKind::Field(..) => match ty.kind() {
+                ty::Adt(def, _) if def.repr.packed() => {
+                    match tcx.layout_raw(param_env.and(p.ty)) {
```

Comment:
I'm wondering whether we can actually get type layout here -- I guess so.

Reply:
(I was thinking about possible cycles.)

Reply:
So, first off, I think we should be calling [`tcx.layout_of`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html#method.layout_of), probably?

Except, maybe not, because that executes in "reveal all" mode, and I'm pretty sure we don't want that.

This is making me wonder if we should not try to handle the 'alignment 1' rule for closure capture.

Reply:
I'm not too sure about this. I essentially looked for code that triggers the repr-packed warning today. I think this was from `rustc_mir`. I was worried about if we will have the information available as well, but it worked.

Reply:
Yeah. It looks like it came from [is_disaligned](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_mir/util/alignment/fn.is_disaligned.html). About the only cycle I can imagine is if you had a packed struct with a field whose type included a constant expression that used that same packed struct, or something like that.  I think it's ok.